docs/design/factories.md

```diff
@@ -0,0 +1,439 @@
+}
+
+func (m *myResource) Create(ctx context.Context, req framework.CreateResourceRequest, resp framework.CreateResourceResponse) {
+  var readResult string
+  err := m.readResult.As(&readResult)
+  if err != nil {
+    panic(err)
+  }
+  // make an API call here using readResult
+}
+```
+
+This code could _sometimes_ work if we're not careful about always generating a
+new `myResource` for each RPC call we handle. But if the inner workings of the
+SDK or Terraform's graph change in any way, it's likely to break this code,
+which may not be obvious to provider developers. We can mitigate this by
+requiring provider developers to register a function with the provider, not a
+value, and always calling the function to get a fresh value at the beginning of
+every RPC call, though provider developers may not understand that we're doing
+that.
+
+Second, there exists a certain kind of state that it's very reasonable for
+providers to want to have available to all RPC calls for every instance of
+their resources. This state usually is _used_ by RPC calls, not _created_ by
+it. An example of this we see a lot in the wild is a mutex that constrains the
+number of requests that can be made in parallel, to not provoke API rate
+limiting. Currently, the only way to keep this state is to register it as
+global mutable state. This is problematic in testing scenarios, as all provider
+servers will need to share that same state; it's not just one server's resource
+instances, it's all the resource instances for all the servers created by any
+concurrently-running tests.
+
+We have the option of surfacing this distinction explicitly, allowing providers
+to store this state that should be shared among all instances of a resource
+type. We could define the resource type and the resource instance as separate
+Go types:
+
+```go
+// in the framework
+type ResourceType interface {
+  GetSchema() *tfprotov5.Schema
+  NewResource(p framework.Provider) Resource
+}
+
+type Resource interface {
+  Create(ctx context.Context, *CreateResourceRequest, *CreateResourceResponse)
+  Read(ctx context.Context, *ReadResourceRequest, *ReadResourceResponse)
+  Update(ctx context.Context, *UpdateResourceRequest, *UpdateResourceResponse)
+  Delete(ctx context.Context, *DeleteResourceRequest, *DeleteResourceResponse)
+}
+```
+
+This allows provider developers to define their resource types and thread
+through the state shared by all instances of the resource:
+
+```go
+// in the provider
+type myResourceType struct {
+  reqMutex sync.Mutex
+}
+
+func (m *myResourceType) NewResource(p framework.Provider) framework.Resource{
+  return &myResource{
+    reqMutex: &m.reqMutex,
+    client: p.(*Client),
+  }
+}
+
+func (m *myResourceType) GetSchema() *tfprotov5.Schema {
+  return &tfprotov5.Schema{
+    // hard-code schema here
+  }
+}
+
+type myResource struct {
+  reqMutex sync.Mutex
+  client *Client
+}
+
+func (m *myResource) Create(ctx, req, resp) {
```

Comment:
Since this is in the provider:

```suggestion
func (m *myResource) Create(ctx context.Context, req *framework.CreateResourceRequest, resp *framework.CreateResourceResponse) {
```